iceoryx_dust/vocabulary/include/iox/span.hpp

```diff
@@ -0,0 +1,427 @@
+
+template <typename T>
+struct is_span : std::false_type
+{
+};
+
+template <typename T, uint64_t Extent>
+struct is_span<span<T, Extent>> : std::true_type
+{
+};
+
+template <typename T>
+using is_not_span_t = iox::negation<is_span<std::decay_t<T>>>;
+
+template <typename Container, typename T>
+using container_has_convertible_data_t =
+    is_convertible_t<std::remove_pointer_t<decltype(iox::data(std::declval<Container>()))>, T>;
+
+template <typename Container>
+using is_integral_sized_container_t = std::is_integral<decltype(iox::size(std::declval<Container>()))>;
+
+template <typename From, uint64_t FromExtent, typename To, uint64_t ToExtent>
+using enable_if_conversion_allowed_t =
+    std::enable_if_t<(ToExtent == DYNAMIC_EXTENT || ToExtent == FromExtent) && is_convertible_t<From, To>::value>;
+
+// SFINAE Verify it the passed array can be converted to a span<T>.
+template <typename Array, typename T, uint64_t Extent>
+using enable_if_compatible_array_t =
+    std::enable_if_t<(Extent == DYNAMIC_EXTENT || Extent == detail::extent_t<Array>::value)
+                     && container_has_convertible_data_t<Array, T>::value>;
+
+// SFINAE Verify it the passed container can be converted to a span<T>.
+template <typename Container, typename T>
+using is_compatible_container_t = conjunction<is_not_span_t<Container>,
+                                              iox::is_not_iox_array_t<Container>,
+                                              iox::is_not_c_array_t<Container>,
+                                              container_has_convertible_data_t<Container, T>,
+                                              is_integral_sized_container_t<Container>>;
+
+template <typename Container, typename T>
+using enable_if_compatible_container_t = std::enable_if_t<is_compatible_container_t<Container, T>::value>;
+
+template <typename Container, typename T, uint64_t Extent>
+using enable_if_compatible_dynamic_container_t =
+    std::enable_if_t<is_compatible_container_t<Container, T>::value && Extent == DYNAMIC_EXTENT>;
+
+template <uint64_t Extent>
+class span_storage
+{
+  public:
+    constexpr explicit span_storage(uint64_t) noexcept
+    {
+    }
+    constexpr uint64_t size() const noexcept
+    {
+        return Extent;
+    }
+};
+
+template <>
+struct span_storage<DYNAMIC_EXTENT>
+{
+    constexpr explicit span_storage(uint64_t size) noexcept
+        : m_size(size)
+    {
+    }
+    constexpr uint64_t size() const noexcept
+    {
+        return m_size;
+    }
+
+    span_storage(const span_storage& other) noexcept = default;
+    span_storage& operator=(const span_storage& other) noexcept = default;
+
+    span_storage(span_storage&& other) noexcept
+    {
+        *this = std::move(other);
+    }
+    span_storage& operator=(span_storage&& other) noexcept
+    {
+        if (this != &other)
+        {
+            m_size = other.m_size;
+            other.m_size = 0;
+        }
+        return *this;
+    }
```

Comment:
Just thinking out loud. We can keep it as is but is it really necessary to manually implement the move c'tor and move assignment operator? Since the `span` does not do resource management, the default move c'tor and assignment operators would also do the job and mirror the non `DYNAMIC_EXTENT` behavior.

Btw, it also seems to be the behavior of `std::span`. I tried out this
```cpp
#include <iostream>
#include <span>
#include <vector>

void print(std::span<int, std::dynamic_extent>&& s) {
    std::cout << "====" << std::endl;
    for(auto i: s) {
        std::cout << i << " ";
    }
    std::cout << std::endl;
}

int main() {
    std::vector<int> v {1, 1, 2, 3, 5, 8};
    std::span<int, std::dynamic_extent> s{v};

    print(std::move(s));
    print(std::move(s));

    return EXIT_SUCCESS;
}
```
and got two times `1 1 2 3 5 8`.

Reply:
@elBoberido Good pointed. They are defaulted now.